File: packages/python/tests/test_pin/transport_contract.py
"""The one canned :class:`HttpClient` double, plus the transport conformance table.

Every adapter behind :class:`~ghagen.pin.github.HttpClient` — the production
:class:`~ghagen.pin.github.UrllibTransport` and the canned
:class:`FakeTransport` alike — must satisfy the same table, so a double cannot
drift into a shape the real adapter is unable to produce.

Two builders construct canned responses: :func:`canned` encodes a JSON value,
:func:`canned_raw` takes a body verbatim (a malformed 200, a truncated payload).

Rows 8-12 put the adapter in front of a **raw socket**, not a request-handling
server: ``http.server`` always frames a well-formed response, so it cannot
express "peer closes without answering" (row 9), "declared
``Content-Length: 100``, delivered 5 bytes" (row 11), or "trickle a byte at a
time forever" (row 12).  Each scenario gets its own listening socket and its
own daemon thread — a scenario whose handler deliberately never returns (row
10) would otherwise block a shared accept loop and deadlock the next one.
"""

from __future__ import annotations

import json
import socket
import threading
import time
from collections.abc import Callable, Iterator, Mapping, Sequence
from contextlib import contextmanager
from dataclasses import dataclass
from typing import Any

import pytest

from ghagen.pin.github import HttpClient, Response, TransportError

# The deadline the adapter under test is built with for the response rows.
# Generous: these rows are about what comes back, not about timing.
RESPONSE_DEADLINE_SECONDS = 5.0

# Row 10's deadline, and the sole caller of the adapters' deadline argument.
FAILURE_DEADLINE_SECONDS = 0.25

# Upper bound on a failure row's wall clock, at 10x the deadline.  No lower
# bound is asserted: a "did it really wait?" check flakes on a fast machine and
# adds nothing the ceiling and the raised type do not already establish.
FAILURE_CEILING_SECONDS = 2.5


# -- the one canned double -------------------------------------------------


def canned(
    obj: object,
    *,
    status: int = 200,
    reason: str = "",
    headers: Mapping[str, str] | None = None,
) -> Response:
    """Build a canned :class:`Response` whose body is *obj* encoded as JSON."""
    return canned_raw(
        json.dumps(obj).encode(), status=status, reason=reason, headers=headers
    )


def canned_raw(
    body: bytes | str,
    *,
    status: int = 200,
    reason: str = "",
    headers: Mapping[str, str] | None = None,
) -> Response:
    """Build a canned :class:`Response` around a verbatim body."""
    return Response(
        status=status,
        body=body.encode() if isinstance(body, str) else body,
        reason=reason,
        headers=dict(headers or {}),
    )


class FakeTransport:
    """Canned :class:`~ghagen.pin.github.HttpClient` keyed by URL substring.

    Each entry maps a URL *substring* to a :class:`Response`, a list of
    responses (consumed in order, for pagination), or an exception to raise.
    Unmatched URLs return a 404.  Requested tokens are recorded on
    :attr:`tokens`, requested URLs on :attr:`calls`.
    """

    def __init__(self, responses: Mapping[str, object] | None = None) -> None:
        self._responses = dict(responses or {})
        self.calls: list[str] = []
        self.tokens: list[str | None] = []

    def get(self, url: str, *, token: str | None = None) -> Response:
        self.calls.append(url)
        self.tokens.append(token)
        for pattern, value in self._responses.items():
            if pattern in url:
                if isinstance(value, list):
                    return value.pop(0)
                if isinstance(value, BaseException):
                    raise value
                assert isinstance(value, Response)
                return value
        return Response(status=404, body=b"{}", reason="Not Found")


# -- the conformance table -------------------------------------------------


@dataclass(frozen=True)
class ResponseCase:
    """A row that must produce a returned :class:`Response` (rows 1-7).

    One description serves both kinds of adapter: :meth:`wire` is what a raw
    origin writes to the socket, :meth:`response` is the equivalent canned
    :class:`Response`.  That is what stops the double from drifting.
    """

    name: str
    status: int
    reason: str
    body: bytes
    headers: tuple[tuple[str, str], ...] = ()
    token: str | None = None

    def wire(self) -> bytes:
        """The literal bytes a raw origin writes for this row."""
        head = f"HTTP/1.1 {self.status} {self.reason}\r\n"
        for name, value in self.headers:
            head += f"{name}: {value}\r\n"
        head += f"Content-Length: {len(self.body)}\r\n\r\n"
        return head.encode() + self.body

    def response(self) -> Response:
        """The equivalent canned :class:`Response`."""
        return canned_raw(
            self.body,
            status=self.status,
            reason=self.reason,
            headers=dict(self.headers),
        )


_NEXT_LINK = '<https://api.github.com/repos/o/r/git/refs/tags?page=2>; rel="next"'

RESPONSE_CASES: tuple[ResponseCase, ...] = (
    ResponseCase(
        name="json-200",
        status=200,
        reason="OK",
        body=json.dumps({"object": {"type": "commit", "sha": "a" * 40}}).encode(),
    ),
    ResponseCase(
        name="html-200",
        status=200,
        reason="OK",
        body=b"<html>not json</html>",
    ),
    ResponseCase(
        name="not-found-404",
        status=404,
        reason="Not Found",
        body=b'{"message": "Not Found"}',
    ),
    ResponseCase(
        name="server-error-500",
        status=500,
        reason="Internal Server Error",
        body=b'{"message": "boom"}',
    ),
    ResponseCase(
        name="link-header",
        status=200,
        reason="OK",
        body=b"[]",
        headers=(("Link", _NEXT_LINK),),
    ),
    ResponseCase(
        name="token-supplied",
        status=200,
        reason="OK",
        body=b"{}",
        token="s3cret",
    ),
    ResponseCase(
        name="no-token",
        status=200,
        reason="OK",
        body=b"{}",
    ),
)
"""Rows 1-7 — every adapter, real or canned, must satisfy these."""

FAILURE_CASES: tuple[str, ...] = (
    "connection-refused",
    "abrupt-close",
    "stall-mid-body",
    "truncated-body",
    "dribble-body",
)
"""Rows 8-12 — a canned double satisfies these by construction, so it skips them."""

ALL_CASES: tuple[ResponseCase | str, ...] = (*RESPONSE_CASES, *FAILURE_CASES)


def case_id(case: ResponseCase | str) -> str:
    """Test id for a table row."""
    return case.name if isinstance(case, ResponseCase) else case


# -- the loopback origin ---------------------------------------------------


def _read_request(conn: socket.socket) -> bytes:
    """Read one request head off *conn* (through the blank line)."""
    data = b""
    while b"\r\n\r\n" not in data:
        chunk = conn.recv(4096)
        if not chunk:
            break
        data += chunk
    return data


# A head that promises 100 bytes and is followed by 5, used by rows 10 and 11.
_UNDERDELIVERED_HEAD = b"HTTP/1.1 200 OK\r\nContent-Length: 100\r\n\r\n"
_UNDERDELIVERED_BODY = b"12345"

# Row 12's shape: one body byte every _DRIBBLE_INTERVAL for _DRIBBLE_SECONDS.
# The interval must exceed FAILURE_DEADLINE_SECONDS/2 so that a *per-socket*
# timeout of one deadline never fires — which is exactly what makes the row
# distinguish a wall-clock deadline from a per-operation one.
_DRIBBLE_INTERVAL = 0.2
_DRIBBLE_SECONDS = 10.0
_DRIBBLE_LENGTH = int(_DRIBBLE_SECONDS / _DRIBBLE_INTERVAL)


class LoopbackOrigin:
    """A raw-socket origin bound to ``127.0.0.1:0``.

    Owns one listening socket and one daemon thread, closed by the context
    manager.  *handler* receives the accepted connection and a
    :class:`threading.Event` that is set during teardown, so a deliberately
    stalling handler can be released rather than joined forever.
    """

    def __init__(
        self, handler: Callable[[socket.socket, threading.Event], None]
    ) -> None:
        self._handler = handler
        self._stop = threading.Event()
        self.requests: list[bytes] = []
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._sock.bind(("127.0.0.1", 0))
        self._sock.listen(4)
        self._sock.settimeout(0.1)
        self._thread = threading.Thread(target=self._serve, daemon=True)

    @property
    def url(self) -> str:
        """A URL this origin will answer."""
        host, port = self._sock.getsockname()
        return f"http://{host}:{port}/contract"

    def observed_auth(self) -> str | None:
        """The ``Authorization`` header value the origin saw, if any."""
        return _auth_header(self.requests[-1] if self.requests else b"")

    def _serve(self) -> None:
        while not self._stop.is_set():
            try:
                conn, _ = self._sock.accept()
            except (TimeoutError, OSError):
                continue
            with conn:
                try:
                    self.requests.append(_read_request(conn))
                    self._handler(conn, self._stop)
                except OSError:  # the client hung up first; nothing to report
                    pass

    def __enter__(self) -> LoopbackOrigin:
        self._thread.start()
        return self

    def __exit__(self, *exc: object) -> None:
        self._stop.set()
        self._thread.join(timeout=FAILURE_CEILING_SECONDS)
        self._sock.close()


def _auth_header(request: bytes) -> str | None:
    """Extract the ``Authorization`` value from a raw request head."""
    for line in request.split(b"\r\n"):
        name, _, value = line.partition(b":")
        if name.strip().lower() == b"authorization":
            return value.strip().decode()
    return None


def _responder(wire: bytes) -> Callable[[socket.socket, threading.Event], None]:
    def handler(conn: socket.socket, _stop: threading.Event) -> None:
        conn.sendall(wire)

    return handler


def _abrupt_close(conn: socket.socket, _stop: threading.Event) -> None:
    """Row 9 — accept the request, then close without writing a response."""
    return None


def _stall_mid_body(conn: socket.socket, stop: threading.Event) -> None:
    """Row 10 — send a head and part of the body, then hold the connection."""
    conn.sendall(_UNDERDELIVERED_HEAD + _UNDERDELIVERED_BODY)
    stop.wait()


def _truncated_body(conn: socket.socket, _stop: threading.Event) -> None:
    """Row 11 — declare 100 bytes, deliver 5, close."""
    conn.sendall(_UNDERDELIVERED_HEAD + _UNDERDELIVERED_BODY)


def _dribble_body(conn: socket.socket, stop: threading.Event) -> None:
    """Row 12 — send one body byte every 200ms for 10s.

    The row rows 10 and 11 cannot express.  A *stalled* peer is caught by a
    per-socket-operation timeout too, because no operation completes; a peer
    that keeps trickling resets that timeout forever, so only an adapter
    honouring a genuine **wall-clock** deadline aborts.  ``AbortSignal.timeout``
    is one; ``urlopen(timeout=)`` alone is not.
    """
    conn.sendall(b"HTTP/1.1 200 OK\r\nContent-Length: %d\r\n\r\n" % _DRIBBLE_LENGTH)
    for _ in range(_DRIBBLE_LENGTH):
        if stop.wait(_DRIBBLE_INTERVAL):
            return
        conn.sendall(b"x")


_FAILURE_HANDLERS: dict[
    str, Callable[[socket.socket, threading.Event], None] | None
] = {
    "connection-refused": None,  # served by a port nobody listens on
    "abrupt-close": _abrupt_close,
    "stall-mid-body": _stall_mid_body,
    "truncated-body": _truncated_body,
    "dribble-body": _dribble_body,
}


@contextmanager
def _closed_port() -> Iterator[str]:
    """Yield a URL on a loopback port that nothing is listening on."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind(("127.0.0.1", 0))
    host, port = sock.getsockname()
    sock.close()
    yield f"http://{host}:{port}/contract"


# -- adapter factories -----------------------------------------------------


@dataclass(frozen=True)
class Bound:
    """An adapter bound to one scenario: the client, the URL, the origin's view."""

    client: HttpClient
    url: str
    observed_auth: Callable[[], str | None]


MakeAdapter = Callable[[Any, float], Any]
"""``make_adapter(case, deadline)`` -> context manager yielding a :class:`Bound`."""


def loopback_adapter(build: Callable[[float], HttpClient]) -> MakeAdapter:
    """Run a *real* adapter against a raw loopback origin.

    ``build(deadline)`` constructs the adapter under test with that wall-clock
    deadline in seconds — the deadline argument's sole caller.
    """

    @contextmanager
    def make(case: ResponseCase | str, deadline: float) -> Iterator[Bound]:
        if isinstance(case, ResponseCase):
            handler = _responder(case.wire())
        elif _FAILURE_HANDLERS[case] is None:
            with _closed_port() as url:
                yield Bound(build(deadline), url, lambda: None)
            return
        else:
            handler = _FAILURE_HANDLERS[case]  # type: ignore[assignment]
        with LoopbackOrigin(handler) as origin:
            yield Bound(build(deadline), origin.url, origin.observed_auth)

    return make


def canned_adapter() -> MakeAdapter:
    """Run :class:`FakeTransport` against the same table (response rows only)."""

    @contextmanager
    def make(case: ResponseCase | str, _deadline: float) -> Iterator[Bound]:
        assert isinstance(case, ResponseCase), (
            "a canned double satisfies rows 8-11 by construction; it runs rows 1-7"
        )
        transport = FakeTransport({"/contract": case.response()})

        def observed_auth() -> str | None:
            token = transport.tokens[-1] if transport.tokens else None
            return f"Bearer {token}" if token else None

        yield Bound(transport, "https://origin.test/contract", observed_auth)

    return make


# -- the runner ------------------------------------------------------------


def assert_transport_contract(
    make_adapter: MakeAdapter,
    *,
    cases: Sequence[ResponseCase | str] = ALL_CASES,
) -> None:
    """Assert that *make_adapter*'s adapter satisfies the conformance table.

    Defaults to the full table; the test modules pass one row at a time so a
    failure names the row.
    """
    for case in cases:
        if isinstance(case, ResponseCase):
            _assert_response_case(make_adapter, case)
        else:
            _assert_failure_case(make_adapter, case)


def _assert_response_case(make_adapter: MakeAdapter, case: ResponseCase) -> None:
    with make_adapter(case, RESPONSE_DEADLINE_SECONDS) as bound:
        resp = bound.client.get(bound.url, token=case.token)
        observed = bound.observed_auth()

    assert resp.status == case.status
    assert resp.body == case.body
    assert resp.reason == case.reason
    for name, value in case.headers:
        assert resp.header(name.lower()) == value
        assert resp.header(name.upper()) == value
    assert observed == (f"Bearer {case.token}" if case.token else None)


def _assert_failure_case(make_adapter: MakeAdapter, name: str) -> None:
    with make_adapter(name, FAILURE_DEADLINE_SECONDS) as bound:
        started = time.monotonic()
        with pytest.raises(TransportError):
            bound.client.get(bound.url)
        elapsed = time.monotonic() - started

    assert elapsed < FAILURE_CEILING_SECONDS, (
        f"{name}: took {elapsed:.2f}s, deadline was {FAILURE_DEADLINE_SECONDS}s"
    )

File: packages/typescript/src/pin/transport-contract.ts
/**
 * The one canned `HttpClient` double, plus the transport conformance table.
 *
 * Every adapter behind `HttpClient` — the production `FetchTransport` and the
 * canned `FakeTransport` alike — must satisfy the same table, so a double
 * cannot drift into a shape the real adapter is unable to produce. The Python
 * peer (`tests/test_pin/transport_contract.py`) runs the identical rows.
 *
 * Two builders construct canned responses: `canned()` encodes a JSON value,
 * `cannedRaw()` takes a body verbatim (a malformed 200, a truncated payload).
 *
 * Rows 9-14 put the adapter in front of a **raw socket**, not a
 * request-handling server: `node:http`'s `createServer` always frames a
 * well-formed response, so it cannot express "peer closes without answering"
 * (row 10), "declared `Content-Length: 100`, delivered 5 bytes" (row 12),
 * "trickle a body byte at a time forever" (row 13), or "trickle a *head*
 * byte at a time, forever, never completing it" (row 14).
 *
 * Test-only; excluded from the build in `tsconfig.json` beside
 * `src/integration/test-utils.ts`.
 */

import { expect } from "vitest";
import { createServer, type AddressInfo, type Server, type Socket } from "node:net";
import { HttpResponse, TransportError, type HttpClient, type RequestOptions } from "./github.js";

/** The deadline the adapter under test is built with for the response rows. */
export const RESPONSE_DEADLINE_MS = 5_000;

/** Row 11's deadline, and the sole caller of the adapters' deadline argument. */
export const FAILURE_DEADLINE_MS = 250;

/**
 * Upper bound on a failure row's wall clock, at 10x the deadline. No lower
 * bound is asserted: a "did it really wait?" check flakes on a fast machine and
 * adds nothing the ceiling and the rejected type do not already establish.
 */
export const FAILURE_CEILING_MS = 2_500;

// ---- the one canned double ----

/** Options shared by both canned-response builders. */
export interface CannedOptions {
  status?: number;
  /** Python's `reason`; the name stays TS-idiomatic. */
  statusText?: string;
  headers?: Record<string, string>;
}

/** Build a canned `HttpResponse` whose body is `value` encoded as JSON. */
export function canned(value: unknown, options: CannedOptions = {}): HttpResponse {
  return cannedRaw(JSON.stringify(value), options);
}

/** Build a canned `HttpResponse` around a verbatim body. */
export function cannedRaw(body: string, options: CannedOptions = {}): HttpResponse {
  return new HttpResponse(
    options.status ?? 200,
    body,
    options.statusText ?? "",
    options.headers ?? {},
  );
}

/** What a `FakeTransport` entry may map a URL substring to. */
export type Canned = HttpResponse | HttpResponse[] | Error;

/**
 * Canned `HttpClient` keyed by URL substring.
 *
 * Each entry maps a URL substring to a response, a list of responses (consumed
 * in order, for pagination), or an error to throw. Unmatched URLs return a 404.
 * Requested tokens are recorded on `tokens`, requested URLs on `calls`.
 */
export class FakeTransport implements HttpClient {
  readonly calls: string[] = [];
  readonly tokens: Array<string | undefined> = [];

  constructor(private readonly responses: Record<string, Canned> = {}) {}

  async get(url: string, options: RequestOptions = {}): Promise<HttpResponse> {
    this.calls.push(url);
    this.tokens.push(options.token);
    for (const [pattern, value] of Object.entries(this.responses)) {
      if (url.includes(pattern)) {
        if (Array.isArray(value)) {
          return value.shift()!;
        }
        if (value instanceof Error) {
          throw value;
        }
        return value;
      }
    }
    return new HttpResponse(404, "{}", "Not Found");
  }
}

// ---- the conformance table ----

/**
 * A row that must produce a resolved `HttpResponse` (rows 1-8).
 *
 * One description serves both kinds of adapter: `wire()` is what a raw origin
 * writes to the socket, `response()` is the equivalent canned `HttpResponse`.
 * That is what stops the double from drifting.
 */
export class ResponseCase {
  constructor(
    readonly name: string,
    readonly status: number,
    readonly statusText: string,
    readonly body: string,
    readonly headers: ReadonlyArray<readonly [string, string]> = [],
    readonly token?: string,
  ) {}

  /** The literal bytes a raw origin writes for this row. */
  wire(): string {
    let head = `HTTP/1.1 ${this.status} ${this.statusText}\r\n`;
    for (const [name, value] of this.headers) {
      head += `${name}: ${value}\r\n`;
    }
    head += `Content-Length: ${Buffer.byteLength(this.body)}\r\n\r\n`;
    return head + this.body;
  }

  /** The equivalent canned `HttpResponse`. */
  response(): HttpResponse {
    return cannedRaw(this.body, {
      status: this.status,
      statusText: this.statusText,
      headers: Object.fromEntries(this.headers),
    });
  }
}

const NEXT_LINK = '<https://api.github.com/repos/o/r/git/refs/tags?page=2>; rel="next"';

/** Rows 1-8 — every adapter, real or canned, must satisfy these. */
export const RESPONSE_CASES: readonly ResponseCase[] = [
  new ResponseCase(
    "json-200",
    200,
    "OK",
    JSON.stringify({ object: { type: "commit", sha: "a".repeat(40) } }),
  ),
  new ResponseCase("html-200", 200, "OK", "<html>not json</html>"),
  new ResponseCase("not-found-404", 404, "Not Found", '{"message": "Not Found"}'),
  new ResponseCase("server-error-500", 500, "Internal Server Error", '{"message": "boom"}'),
  new ResponseCase("link-header", 200, "OK", "[]", [["Link", NEXT_LINK]]),
  new ResponseCase("token-supplied", 200, "OK", "{}", [], "s3cret"),
  new ResponseCase("no-token", 200, "OK", "{}"),
  // Row 8 — a paginated page that parses as JSON but is not an array. The
  // transport resolves this like any other 200; it is `GitHubClient.listTags`
  // (`refNames`, `github.ts`) that must turn it into `ResolveError` rather than
  // truncating the tag list or crashing outright. Binding the raw bytes here
  // keeps both adapters delivering the identical malformed body so that
  // per-port shape-check, which is tested separately, sees the same input.
  new ResponseCase("non-array-page", 200, "OK", '{"message": "not an array"}'),
];

/** Rows 9-14 — a canned double satisfies these by construction, so it skips them. */
export const FAILURE_CASES: readonly string[] = [
  "connection-refused",
  "abrupt-close",
  "stall-mid-body",
  "truncated-body",
  "dribble-body",
  "dribble-head",
];

export const ALL_CASES: ReadonlyArray<ResponseCase | string> = [
  ...RESPONSE_CASES,
  ...FAILURE_CASES,
];

/** Test id for a table row. */
export function caseId(testCase: ResponseCase | string): string {
  return typeof testCase === "string" ? testCase : testCase.name;
}

// ---- the loopback origin ----

/** A head that promises 100 bytes and is followed by 5, used by rows 11 and 12. */
const UNDERDELIVERED = "HTTP/1.1 200 OK\r\nContent-Length: 100\r\n\r\n12345";

/**
 * Row 13's shape: one body byte every `DRIBBLE_INTERVAL_MS` for 10s.
 *
 * The interval must exceed `FAILURE_DEADLINE_MS / 2` so that a *per-socket*
 * timeout of one deadline never fires — which is exactly what makes the row
 * distinguish a wall-clock deadline from a per-operation one.
 */
const DRIBBLE_INTERVAL_MS = 200;
const DRIBBLE_LENGTH = 50;

type Handler = (socket: Socket) => void;

/**
 * A raw-socket origin bound to `127.0.0.1:0`.
 *
 * Teardown destroys every tracked socket before `close()`: a connection with an
 * in-flight request the handler never answers (row 11) blocks the `close`
 * callback indefinitely otherwise.
 */
export class LoopbackOrigin {
  private readonly server: Server;
  private readonly sockets = new Set<Socket>();
  readonly requests: string[] = [];
  private port = 0;

  constructor(handler: Handler) {
    this.server = createServer((socket) => {
      this.sockets.add(socket);
      socket.on("close", () => this.sockets.delete(socket));
      socket.on("error", () => {}); // the client hanging up is not a failure
      let head = "";
      const onData = (chunk: Buffer): void => {
        head += chunk.toString("latin1");
        if (head.includes("\r\n\r\n")) {
          socket.off("data", onData);
          this.requests.push(head);
          handler(socket);
        }
      };
      socket.on("data", onData);
    });
  }

  async start(): Promise<void> {
    await new Promise<void>((resolve) => this.server.listen(0, "127.0.0.1", resolve));
    this.port = (this.server.address() as AddressInfo).port;
  }

  /** A URL this origin will answer. */
  get url(): string {
    return `http://127.0.0.1:${this.port}/contract`;
  }

  /** The `Authorization` header value the origin saw, if any. */
  observedAuth(): string | null {
    return authHeader(this.requests[this.requests.length - 1] ?? "");
  }

  async stop(): Promise<void> {
    for (const socket of this.sockets) {
      socket.destroy();
    }
    await new Promise<void>((resolve) => this.server.close(() => resolve()));
  }
}

/** Extract the `Authorization` value from a raw request head. */
function authHeader(request: string): string | null {
  for (const line of request.split("\r\n")) {
    const idx = line.indexOf(":");
    if (idx > 0 && line.slice(0, idx).trim().toLowerCase() === "authorization") {
      return line.slice(idx + 1).trim();
    }
  }
  return null;
}

const FAILURE_HANDLERS: Record<string, Handler | null> = {
  // Served by a port nobody listens on.
  "connection-refused": null,
  /** Row 10 — accept the request, then close without writing a response. */
  "abrupt-close": (socket) => socket.end(),
  /** Row 11 — send a head and part of the body, then hold the connection. */
  "stall-mid-body": (socket) => {
    socket.write(UNDERDELIVERED);
  },
  /** Row 12 — declare 100 bytes, deliver 5, close. */
  "truncated-body": (socket) => {
    socket.write(UNDERDELIVERED);
    socket.end();
  },
  /**
   * Row 13 — send one body byte every 200ms for 10s.
   *
   * The row rows 11 and 12 cannot express. A *stalled* peer is caught by a
   * per-socket-operation timeout too, because no operation completes; a peer
   * that keeps trickling resets that timeout forever, so only an adapter
   * honouring a genuine **wall-clock** deadline aborts. `AbortSignal.timeout`
   * is one; `urlopen(timeout=)` alone is not.
   */
  "dribble-body": (socket) => {
    socket.write(`HTTP/1.1 200 OK\r\nContent-Length: ${DRIBBLE_LENGTH}\r\n\r\n`);
    let sent = 0;
    const timer = setInterval(() => {
      if (sent++ >= DRIBBLE_LENGTH || socket.destroyed) {
        clearInterval(timer);
        return;
      }
      socket.write("x");
    }, DRIBBLE_INTERVAL_MS);
    timer.unref();
    socket.on("close", () => clearInterval(timer));
  },
  /**
   * Row 14 — send one head byte every 200ms for 10s, never completing it.
   *
   * `dribble-body` (row 13) proves a wall-clock deadline once the head is
   * already in hand; this row proves the same thing about the head itself —
   * no status line, no headers, ever, just a trickle that a per-operation
   * timeout renews forever. An adapter that only starts its deadline once it
   * sees body bytes — the exact gap a fix for row 13 alone could leave open —
   * hangs on this row instead of failing it.
   */
  "dribble-head": (socket) => {
    let sent = 0;
    const timer = setInterval(() => {
      if (sent++ >= DRIBBLE_LENGTH || socket.destroyed) {
        clearInterval(timer);
        return;
      }
      socket.write("H");
    }, DRIBBLE_INTERVAL_MS);
    timer.unref();
    socket.on("close", () => clearInterval(timer));
  },
};

/** A URL on a loopback port that nothing is listening on. */
async function closedPortUrl(): Promise<string> {
  const origin = new LoopbackOrigin(() => {});
  await origin.start();
  const { url } = origin;
  await origin.stop();
  return url;
}

// ---- adapter factories ----

/** An adapter bound to one scenario: the client, the URL, the origin's view. */
export interface Bound {
  client: HttpClient;
  url: string;
  observedAuth(): string | null;
}

/** `makeAdapter(case, deadlineMs, body)` runs `body` against a bound adapter. */
export type MakeAdapter = (
  testCase: ResponseCase | string,
  deadlineMs: number,
  body: (bound: Bound) => Promise<void>,
) => Promise<void>;

/**
 * Run a *real* adapter against a raw loopback origin.
 *
 * `build(deadlineMs)` constructs the adapter under test with that wall-clock
 * deadline — the deadline argument's sole caller.
 */
export function loopbackAdapter(build: (deadlineMs: number) => HttpClient): MakeAdapter {
  return async (testCase, deadlineMs, body) => {
    if (typeof testCase === "string" && FAILURE_HANDLERS[testCase] === null) {
      const url = await closedPortUrl();
      await body({ client: build(deadlineMs), url, observedAuth: () => null });
      return;
    }
    const handler =
      typeof testCase === "string" ? FAILURE_HANDLERS[testCase]! : responder(testCase.wire());
    const origin = new LoopbackOrigin(handler);
    await origin.start();
    try {
      await body({
        client: build(deadlineMs),
        url: origin.url,
        observedAuth: () => origin.observedAuth(),
      });
    } finally {
      await origin.stop();
    }
  };
}

function responder(wire: string): Handler {
  return (socket) => {
    socket.write(wire);
  };
}

/** Run `FakeTransport` against the same table (response rows only). */
export function cannedAdapter(): MakeAdapter {
  return async (testCase, _deadlineMs, body) => {
    if (typeof testCase === "string") {
      throw new Error("a canned double satisfies rows 9-12 by construction; it runs rows 1-8");
    }
    const transport = new FakeTransport({ "/contract": testCase.response() });
    await body({
      client: transport,
      url: "https://origin.test/contract",
      observedAuth: () => {
        const token = transport.tokens[transport.tokens.length - 1];
        return token ? `Bearer ${token}` : null;
      },
    });
  };
}

// ---- the runner ----

/**
 * Assert that `makeAdapter`'s adapter satisfies the conformance table.
 *
 * Defaults to the full table; the test module passes one row at a time so a
 * failure names the row.
 */
export async function runTransportContract(
  makeAdapter: MakeAdapter,
  cases: ReadonlyArray<ResponseCase | string> = ALL_CASES,
): Promise<void> {
  for (const testCase of cases) {
    if (typeof testCase === "string") {
      await assertFailureCase(makeAdapter, testCase);
    } else {
      await assertResponseCase(makeAdapter, testCase);
    }
  }
}

async function assertResponseCase(makeAdapter: MakeAdapter, testCase: ResponseCase): Promise<void> {
  await makeAdapter(testCase, RESPONSE_DEADLINE_MS, async (bound) => {
    const resp = await bound.client.get(bound.url, { token: testCase.token });
    expect(resp.status).toBe(testCase.status);
    expect(resp.body).toBe(testCase.body);
    expect(resp.statusText).toBe(testCase.statusText);
    for (const [name, value] of testCase.headers) {
      expect(resp.header(name.toLowerCase())).toBe(value);
      expect(resp.header(name.toUpperCase())).toBe(value);
    }
    expect(bound.observedAuth()).toBe(testCase.token ? `Bearer ${testCase.token}` : null);
  });
}

async function assertFailureCase(makeAdapter: MakeAdapter, name: string): Promise<void> {
  await makeAdapter(name, FAILURE_DEADLINE_MS, async (bound) => {
    const started = Date.now();
    await expect(bound.client.get(bound.url)).rejects.toBeInstanceOf(TransportError);
    const elapsed = Date.now() - started;
    expect(elapsed, `${name}: deadline was ${FAILURE_DEADLINE_MS}ms`).toBeLessThan(
      FAILURE_CEILING_MS,
    );
  });
}
